src/nn/transformer.rs

```diff
@@ -0,0 +1,1097 @@
+use crate::prelude::*;
+use rand::Rng;
+
+/// A multi-head attention layer.
+///
+/// # Generics
+/// - `M` The embedding size of token vectors from decoder.
+/// - `N` The embedding size of token vectors from encoder.
+/// - `K` The size of the keys in self attention.
+/// - `V` The size of the values.
+/// - `H` The number of attention heads.
+///
+/// # Examples
+/// `MultiHeadAttention<8, 10, 10, 10, 2>` is an attention layer with 2 heads and 10 token, key and value dims.
+/// TODO: Doctests fail for some reason
+#[derive(Debug, Clone, Default)]
+pub struct MultiHeadAttention<
+    const M: usize,
+    const N: usize,
+    const K: usize,
+    const V: usize,
+    const H: usize,
+> {
+    w_q: Linear<N, K>,
+    w_k: Linear<M, K>,
+    w_v: Linear<M, V>,
+    w_o: Linear<V, M>,
```

Comment:
I think these all need to be multiplied by number of heads (`H`), right? Otherwise there's only just 1 head?

Reply:
No, there are H heads but each is M / H dims, so all heads combined are M dims. The projection is combined for all heads, and the head splitting only happens after. Take a look at the minGPT implementation: https://github.com/karpathy/minGPT/blob/master/mingpt/model.py